### PR TITLE
Hardware key serial number attestation

### PR DESCRIPTION
**Pull request by Joerger**
Changelog: Add option to validate hardware key serial numbers with hardware key support.

During hardware key login, attest the user's hardware key serial number to confirm that the hardware key is trusted by checking the user's traits.

Depends on https://github.com/gravitational/teleport/pull/37727 for proto/type changes

e PR - https://github.com/gravitational/teleport.e/pull/3358

### CODE REVIEW

File: api/types/authentication.go [hunk 1]
```diff
@@ -688,8 +688,10 @@ func (c *AuthPreferenceV2) CheckAndSetDefaults() error {
...
 
 	if hk, err := c.GetHardwareKey(); err == nil {
-		if err := keys.PIVSlot(hk.PIVSlot).Validate(); err != nil {
-			return trace.Wrap(err)
+		if hk.PIVSlot != "" {
+			if err := keys.PIVSlot(hk.PIVSlot).Validate(); err != nil {
+				return trace.Wrap(err)
+			}
 		}
 	}
```
Comment (tigrato) [inline api/types/authentication.go:696]:
```go

	if hk, err := c.GetHardwareKey(); err == nil && hk.PIVSlot != "" {
		if err := keys.PIVSlot(hk.PIVSlot).Validate(); err != nil {
			return trace.Wrap(err)
		}
	}


```

File: lib/auth/auth.go [hunk 2]
```diff
@@ -2502,13 +2502,52 @@ func generateCert(a *Server, req certRequest, caType types.CertAuthType) (*proto
...
 			return nil, trace.Wrap(err)
 		}
+
+		if hksn, err := authPref.GetHardwareKeySerialNumberValidation(); err == nil && hksn.Enabled {
+			const defaultSerialNumberTraitName = "hardware_key_serial_numbers"
+			// Note: currently only yubikeys are supported as hardware keys. If we extend
+			// support to more hardware keys, we can add prefixes to serial numbers.
+			// Ex: solokey_12345678 or s_12345678.
+			// We can default to assuming the serial number is a yubikey, so adding this
+			// prefix logic is not needed at this time.
```
Comment (tigrato) [inline lib/auth/auth.go:2528]: What are the implications here if in the feature we add support for other keys? 
Will it cause a breaking change once we extend the support for other keys?

File: lib/auth/auth.go [hunk 3]
```diff
@@ -2502,13 +2502,52 @@ func generateCert(a *Server, req certRequest, caType types.CertAuthType) (*proto
...
+			serialNumberTraitName := hksn.SerialNumberTraitName
+			if serialNumberTraitName == "" {
+				serialNumberTraitName = defaultSerialNumberTraitName
+			}
+
+			// Check that the attested hardware key serial number matches
+			// a serial number in the user's traits, if any are set.
+			registeredSerialNumbers, ok := req.checker.Traits()[serialNumberTraitName]
+			if !ok || len(registeredSerialNumbers) == 0 {
+				log.Debugf("user %q tried to sign in with hardware key support, but has no known hardware keys. A user's known hardware key serial numbers should be set in %q", req.user.GetName(), serialNumberTraitName)
```
Comment (tigrato) [inline lib/auth/auth.go:2538]: it's worth mentioning that it is set in traits

File: lib/auth/auth.go [hunk 4]
```diff
@@ -2502,13 +2502,52 @@ func generateCert(a *Server, req certRequest, caType types.CertAuthType) (*proto
...
+				return nil, trace.BadParameter("cannot generate certs for user with no known hardware keys")
+			}
+
+			attestatedSerialNumber := strconv.Itoa(int(attestationData.SerialNumber))
+			// serial number traits can be a comma seperated list, or a list of comma separated lists.
```
Comment (tigrato) [inline lib/auth/auth.go:2543]: if the format is `"12345678,87654321"`, and the serial number is `87654321`, it won't match the keys as far as I understood

File: lib/auth/auth.go (hunk 2 above)
Comment (Joerger) [reply lib/auth/auth.go:2528]: No, adding the prefixes can be done backwards compatibly. I'll reword the comment for clarity.

File: lib/auth/auth.go (hunk 4 above)
Comment (Joerger) [reply lib/auth/auth.go:2544]: Thanks for catching this, I had the unit test covering this inversed to expect an error :man_facepalming:

**Review by tigrato — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Comment by Joerger** [comment]
@zmb3 Can I get an `/excludeflake *`, it's just timing out from the number of affected tests

**Pull request merged**